graphql_compiler/tests/test_input_data.py

```diff
@@ -756,6 +756,104 @@ def simple_recurse():
         type_equivalence_hints=None)
 
 
+def traverse_then_recurse():
+    graphql_input = '''{
+        Animal {
+            name @output(out_name: "animal_name")
+            out_Animal_ImportantEvent {
+                ... on Event {
+                    name @output(out_name: "important_event")
+                }
+            }
+            out_Animal_ParentOf @recurse(depth: 2) {
+                name @output(out_name: "ancestor_name")
+            }
+        }
+    }'''
+
+    expected_output_metadata = {
+        'animal_name': OutputMetadata(type=GraphQLString, optional=False),
+        'important_event': OutputMetadata(type=GraphQLString, optional=False),
+        'ancestor_name': OutputMetadata(type=GraphQLString, optional=False),
+    }
+
+    expected_input_metadata = {}
+
+    return CommonTestData(
+        graphql_input=graphql_input,
+        expected_output_metadata=expected_output_metadata,
+        expected_input_metadata=expected_input_metadata,
+        type_equivalence_hints=None)
+
+
+def traverse_in_filter_then_recurse():
+    graphql_input = '''{
+        Animal @filter(op_name: "name_or_alias", value: ["$animal_name_or_alias"]) {
+            name @output(out_name: "animal_name")
+            out_Animal_ImportantEvent {
+                ... on Event {
+                    name @output(out_name: "important_event")
+                }
+            }
+            out_Animal_ParentOf @recurse(depth: 2) {
+                name @output(out_name: "ancestor_name")
+            }
+        }
+    }'''
+
+    expected_output_metadata = {
+        'animal_name': OutputMetadata(type=GraphQLString, optional=False),
+        'important_event': OutputMetadata(type=GraphQLString, optional=False),
+        'ancestor_name': OutputMetadata(type=GraphQLString, optional=False),
+    }
+
+    expected_input_metadata = {
+        'animal_name_or_alias': GraphQLString
+    }
+
+    return CommonTestData(
+        graphql_input=graphql_input,
+        expected_output_metadata=expected_output_metadata,
+        expected_input_metadata=expected_input_metadata,
+        type_equivalence_hints=None)
+
+
+def double_recurse():
```

Comment:
How about `two_consecutive_recurses` to differentiate from a potential double nested recurse?

Reply:
ditto